GMX-Bots/gmx5-detect-frontruns/src/agent.ts

```diff
@@ -0,0 +1,206 @@
+        let victimTokenInAmount = 0;
+        let victimTokenOutAmount = 0;
+        let frontrunnerAddress = "";
+        let frontrunnerProfit = 0;
+        let victimAddressAndTokens: string = "";
+        let frontrunnerAddressAndTokens: string = "";
+
+        //check if there is a transaction that completes the sandwich
+        if (callHistory.has(accountBack + token0 + token1)) {
+          let potentialSandwhichFront: [LogDescription, number, string] = callHistory.get(
+            accountBack + token0 + token1
+          )!;
+          const {
+            account: accountFront,
+            tokenIn: tokenInFront,
+            tokenOut: tokenOutFront,
+            amountIn: amountInFront,
+            amountOut: amountOutFront,
+          } = potentialSandwhichFront[0].args;
+
+          //check if the front transaction trades the same tokens in the opposite direction as the back transaction
+          if (tokenInBack == tokenOutFront && tokenOutBack == tokenInFront) {
+            //find any victim transaction
+            callHistoryKeys.forEach((addressAndTokens) => {
+              let event = callHistory.get(addressAndTokens)!;
+              const {
+                account: accountVictim,
+                tokenIn: tokenInVictim,
+                tokenOut: tokenOutVictim,
+                amountIn: amountInVictim,
+                amountOut: amountOutVictim,
+              } = event[0].args;
+              //check if victim was inside the sandwich
+              if (
+                (event[1] > potentialSandwhichFront[1] &&
+                  event[1] < callHistoryIndex &&
+                  callHistoryIndex > potentialSandwhichFront[1]) ||
+                (event[1] < potentialSandwhichFront[1] &&
+                  event[1] < callHistoryIndex &&
+                  callHistoryIndex < potentialSandwhichFront[1]) ||
+                (event[1] > potentialSandwhichFront[1] &&
+                  event[1] > callHistoryIndex &&
+                  callHistoryIndex < potentialSandwhichFront[1])
+              ) {
+                //check if the victim transaction trades the same tokens in the same direction as the front transaction
+                if (tokenInVictim == tokenInFront && tokenOutVictim == tokenOutFront) {
+                  sandwichAttack = true;
+                  victimAddress = accountVictim;
+                  frontHash = potentialSandwhichFront[2];
+                  victimHash = event[2];
+                  victimTokenIn = tokenInVictim;
+                  victimTokenOut = tokenOutVictim;
+                  victimTokenInAmount = amountInVictim;
+                  victimTokenOutAmount = amountOutVictim;
+                  frontrunnerAddress = accountFront;
+                  frontrunnerProfit = amountOutBack - amountInFront;
+                  victimAddressAndTokens = addressAndTokens;
+                  frontrunnerAddressAndTokens = accountBack + token0 + token1;
+                }
+              }
```

Comment:
Why you decided to store the values then check again if there's a frontRun later. 

It would be more organized to have something like this pattern:
```
if ( victim condition checked) {
// set sandwichAttack to true
// delete both victim and frontrunner from history 
// create finding using the the event args. ( no need to store them somewhere else) 
}
```

Reply:
I did it this way because of certain operations with the cleaner that need to occur before the victim and frontrunner are deleted from history. I tried to change it and move the cleaner to another part, but it was causing errors. I think in this case it might not be worth the time to change, but for next bots I will try to use a more clear pattern like the one you suggested